Add phrase confirmation to domain block dialog

noellabo opened the pull request:
![image](https://user-images.githubusercontent.com/28195220/192143399-90058507-49a9-4808-89b8-6e5d20e7a5f8.png)

Add phrase input to the confirmation dialog to inhibit accidental execution of dangerous operations.

noellabo committed:
Add phrase confirmation to domain block dialog

SouthFox-D commented:
Not everyone understands what "domain name" means (especially in other languages),  so it would be better to use "Please type {domain name} to confirm"?

noellabo committed:
Fix messages, css and missing changes

noellabo commented:
![image](https://user-images.githubusercontent.com/28195220/192246948-c275579b-14e3-4996-aa5f-229f11228617.png)

- I changed the wording. Thanks! @SouthFox-D
- Added a missing support for the part that calls the domain block
- adjusted the style for long labels

shleeable commented:
This is excellent.

noellabo committed:
Add setting to require domain input for domain block

noellabo commented:
![image](https://user-images.githubusercontent.com/28195220/192395366-e08d84b8-1737-4bbd-87cc-6233880cc4ef.png)

Add setting to disable input requests. 

This pull request was created for people who accidentally lost a large number of following and followers. Therefore, the default is enabled.

noellabo committed:
Specify exception for rubocop

ClearlyClaire commented:
Also, I can't believe I'm only thinking about it right now, but there are multiple cases in which this may make harassment situations worse: long domain names, domain names with slurs, and international domain names in scripts you can't read or don't know how to type (sure, you can always copy-paste, but that may be prove troublesome, especially in a stressful situation and/or on mobile)

noellabo commented:
The purpose is to prevent accidental execution, so let's simply ask to type `block`.

noellabo commented:
In the reports of misoperation accidents so far, there is a case where the smartphone is repeatedly tapped in the pocket. In this case, simply adding weights or asking for multiple taps will not help. That's why I think it makes sense to ask for input.

I think making the button red is a good idea.

ClearlyClaire commented:
I think “Please type [context-specific phrase] to confirm” with a short translatable context-specific phrase like in your screenshot would work, yeah. I'd add quotes around it. “block” or “block domain” sound good, for groups it would probably be “delete group”.

noellabo committed:
Change input string and change button color for destructive operations

ClearlyClaire commented:
I will do a proper review later, but this looks good to me!

ClearlyClaire commented on app/javascript/mastodon/containers/domain_container.js:
```diff
@@ -21,6 +22,8 @@ const mapDispatchToProps = (dispatch, { intl }) => ({
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Maybe this should be translatable?
Also I don't find the `...expression` form very readable with a condition.

ClearlyClaire commented on app/javascript/mastodon/containers/status_container.js:
```diff
@@ -226,6 +226,8 @@ const mapDispatchToProps = (dispatch, { intl, contextType }) => ({
       message: <FormattedMessage id='confirmations.domain_block.message' defaultMessage='Are you really, really sure you want to block the entire {domain}? In most cases a few targeted blocks or mutes are sufficient and preferable. You will not see content from that domain in any public timelines or your notifications. Your followers from that domain will be removed.' values={{ domain: <strong>{d ...
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Same wrt. `block` being translatable and `...expression` form.

ClearlyClaire commented on app/javascript/mastodon/features/account_timeline/containers/header_container.js:
```diff
@@ -114,6 +114,8 @@ const mapDispatchToProps = (dispatch, { intl }) => ({
       confirm: intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Same.

ClearlyClaire commented on app/javascript/mastodon/features/status/index.js:
```diff
@@ -396,6 +396,8 @@ class Status extends ImmutablePureComponent {
       confirm: this.props.intl.formatMessage(messages.blockDomainConfirm),
       onConfirm: () => this.props.dispatch(blockDomain(domain)),
+      destructive: true,
+      ...confirmDomainBlock && { passphrase: 'block' },
```
Same.

ClearlyClaire commented on app/javascript/mastodon/features/ui/components/confirmation_modal.js:
```diff
@@ -14,22 +15,41 @@ class ConfirmationModal extends React.PureComponent {
     onSecondary: PropTypes.func,
     closeWhenConfirm: PropTypes.bool,
+    destructive: PropTypes.bool,
+
```
Why the extra blank line?

noellabo replied:
Ow!

ClearlyClaire commented on app/lib/user_settings_decorator.rb:
```diff
@@ -14,32 +14,35 @@ def update(settings)
 
   private
 
+  # rubocop: disable Metrics/AbcSize, Metrics/CyclomaticComplexity, Metrics/PerceivedComplexity
```
I wouldn't enable that. This class definitely needs refactoring. But that's for another PR (e.g. #18368)

noellabo replied:
A commit that avoids the warning will revert.

noellabo committed:
Revert "Specify exception for rubocop"

This reverts commit 0003aa43b501f2af90e4679b53b029549014aa0a.

noellabo committed:
Remove unused code in domain_container

noellabo committed:
Changed passphrase to translatable and modified the code to be more readable

noellabo commented:
- Removed unused code from domain_container
- Passphrases can now be translated
- Conditional judgments that are difficult to understand have been reworded to more general wording
- I removed the code that bypasses the rubocop warnings. These will be resolved in another PR https://github.com/mastodon/mastodon/pull/18368

Gargron commented:
If the point is to prevent accidental clicks, wouldn't this also work with having a checkbox that says "Yes, block {domain}"?

noellabo commented:
In the light of real-life cases, that is not enough.

The problem was
- When the user's judgment is declining
- When multiple taps occur randomly

Well, increasing the checkboxes to 3 might be enough, but I think it would be better to be strict enough to require input.

connorshea commented:
This'd be good to have since I've seen a decent number of people accidentally do this to themselves. A simple checkbox for "Yes, I want to block this whole instance" would work for me, since the main thing is just putting a bit of friction into the process to make sure it's not as easy to shoot yourself in the foot.

mjankowski commented:
Doing old issue triage -- I'm going to close this since it seems like the consensus of the review was "lets just do a standard confirm dialog"; lack of activity in ~2 years; needs rebase.

That said, discussion was not totally conclusive ... so if you want to nudge forward, please rebase/reopen and in theory someone will review again.
